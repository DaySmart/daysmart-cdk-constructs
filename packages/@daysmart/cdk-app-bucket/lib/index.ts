import cdk = require('@aws-cdk/core');
import s3 = require('@aws-cdk/aws-s3');

export class AppBucket extends cdk.Construct {
    constructor(scope: cdk.Construct, id: string, props: any) {
        super(scope, id);

        new s3.Bucket(this, 'Bucket', {
            bucketName: `${props.stage}-${props.appName.toLowerCase()}-website-bucket`,
            publicReadAccess: true,
            versioned: true,
        });
    }
}