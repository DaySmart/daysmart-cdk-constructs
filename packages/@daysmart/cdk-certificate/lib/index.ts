import * as cdk from '@aws-cdk/core';
import acm = require('@aws-cdk/aws-certificatemanager');
import route53 = require('@aws-cdk/aws-route53');

export interface CdkCertificateProps {
    domainName: string;
    stage: string;
    project: string;
}

export class CdkCertificate extends cdk.Construct {

  constructor(scope: cdk.Construct, id: string, props: CdkCertificateProps) {
    super(scope, id);

    const hostedZone = route53.HostedZone.fromLookup(this, 'HostedZone', {
        domainName: props.domainName
    });

    new acm.Certificate(this, 'Certificate', {
        domainName: props.domainName,
        subjectAlternativeNames: [
            `${props.stage}.${props.project}.${props.domainName}`,
            `*.${props.stage}.${props.project}.${props.domainName}`
        ],
        validation: acm.CertificateValidation.fromDns(hostedZone)
    });
  }
}
