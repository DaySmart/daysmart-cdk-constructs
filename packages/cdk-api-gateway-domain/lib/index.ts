import * as cdk from '@aws-cdk/core';
import * as apigw from '@aws-cdk/aws-apigateway';
import * as acm from '@aws-cdk/aws-certificatemanager';
import * as route53 from '@aws-cdk/aws-route53';

export interface CdkApiGatewayDomainProps {
  companyDomainName: string;
  companyHostedZoneId: string;
  project: string;
  baseEnv: string;
  dynamicEnv?: string;
  certificateArn: string;
  restApiId: string;
  basePath: string;
}

export class CdkApiGatewayDomain extends cdk.Construct {

  constructor(scope: cdk.Construct, id: string, props: CdkApiGatewayDomainProps) {
    super(scope, id);

    const api = apigw.RestApi.fromRestApiId(this, "Rest API", `${props.restApiId}`);

    let customDomain: apigw.DomainName;

    if (props.dynamicEnv) {
      customDomain = new apigw.DomainName(this, 'Custom Domain', {
        domainName: `${props.dynamicEnv}-api.${props.baseEnv}.${props.project}.${props.companyDomainName}`,
        certificate: acm.Certificate.fromCertificateArn(this, "Certificate", `${props.certificateArn}`),
        endpointType: apigw.EndpointType.EDGE,
        securityPolicy: apigw.SecurityPolicy.TLS_1_0
      });
    } else {
      customDomain = new apigw.DomainName(this, 'Custom Domain', {
        domainName: (props.baseEnv == "prod") ? `api.${props.project}.${props.companyDomainName}` : `api.${props.baseEnv}.${props.project}.${props.companyDomainName}`,
        certificate: acm.Certificate.fromCertificateArn(this, "Certificate", `${props.certificateArn}`),
        endpointType: apigw.EndpointType.EDGE,
        securityPolicy: apigw.SecurityPolicy.TLS_1_0
      });
    }

    const cloudformationBasePathMapping = new apigw.CfnBasePathMapping(this, "CloudformationBasePathMapping", {
      basePath: `${props.basePath}`,
      domainName: customDomain.domainName,
      restApiId: `${props.restApiId}`,
      stage: `${props.baseEnv}` 
    });

    const cloudformationRoute53Change = new route53.CfnRecordSet(this, "CloudformationAliasRecordRoute53", {
      aliasTarget: {
        dnsName: customDomain.domainNameAliasDomainName,
        evaluateTargetHealth: false,
        hostedZoneId: "Z2FDTNDATAQYW2"
      },
      comment: `CloudFront distribution for ${customDomain.domainName}`,
      hostedZoneId: props.companyHostedZoneId,
      name: customDomain.domainName,
      type: "A"
    });
  }
}
