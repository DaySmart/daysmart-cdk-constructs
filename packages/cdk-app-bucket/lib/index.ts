import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { aws_s3 as s3 } from 'aws-cdk-lib';
import { aws_cloudfront as cloudfront } from 'aws-cdk-lib';
import { aws_iam as iam } from 'aws-cdk-lib';

export interface AppBucketProps {
    stage: string;
    appName: string;
    dynamicEnvName: string;
    projectName: string;
    sharedServicesAccountId?: string;
}
export class AppBucket extends Construct {
    constructor(scope: Construct, id: string, props: AppBucketProps) {
        super(scope, id);

        const bucket = new s3.Bucket(this, 'Bucket', {
            bucketName: `${props.dynamicEnvName}-${props.appName}.${props.stage}.${props.projectName}`,
            versioned: true,
            objectOwnership: s3.ObjectOwnership.BUCKET_OWNER_PREFERRED,
            blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL
        });

        const originAccessIdentity = new cloudfront.OriginAccessIdentity(this, 'OriginAcessIdentity', {
            comment: `OriginAccessIdentity for ${bucket.bucketName}.`
        });

        bucket.grantRead(originAccessIdentity)

        if(props.sharedServicesAccountId) {
            bucket.grantReadWrite(new iam.AccountPrincipal(props.sharedServicesAccountId));
        }

        let output = new cdk.CfnOutput(this, "AppBucket", {
            value: bucket.bucketName
        });

        let output2 = new cdk.CfnOutput(this, "OriginAccessIdentity", {
            value: originAccessIdentity.originAccessIdentityName
        });

        output.overrideLogicalId("AppBucket");
        output2.overrideLogicalId("OriginAccessIdentity");
    }
}