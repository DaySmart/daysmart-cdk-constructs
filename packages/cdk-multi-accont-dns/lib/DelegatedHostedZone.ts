import * as cdk from '@aws-cdk/core';
import * as route53 from '@aws-cdk/aws-route53';

/**
 * Properties for a hosted zone to be delegated
 */
export interface DelegatedHostedZoneProps {
    /**
     * Name of the hosted zone. i.e. example.domain.com
     */
    zoneName: string;
}

/**
 * A Route 53 hosted zone to be delegated from another account
 */
export class DelegatedHostedZone extends cdk.Construct {
    constructor(scope: cdk.Construct, id: string, props: DelegatedHostedZoneProps) {
        super(scope, id);

        const hostedZone = new route53.HostedZone(this, 'HostedZone', {
            zoneName: props.zoneName
        });

        const zoneId = new cdk.CfnOutput(this, 'HostedZoneId', {
            value: hostedZone.hostedZoneId
        });
        zoneId.overrideLogicalId('HostedZoneId');

        if(hostedZone.hostedZoneNameServers) {
            const nameServersList = hostedZone.hostedZoneNameServers as string[];
            const nameServers = new cdk.CfnOutput(this, 'NameServers', {
                value: cdk.Fn.join(',', nameServersList)
            });
            nameServers.overrideLogicalId('NameServers');
        }
    }

}