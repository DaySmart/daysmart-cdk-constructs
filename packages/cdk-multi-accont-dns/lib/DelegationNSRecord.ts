import * as cdk from 'aws-cdk-lib';
import * as route53 from 'aws-cdk-lib/aws-route53';
import { Construct } from 'constructs';

/**
 * Properties for an NS record to delegate a hosted zone
 */
export interface DelegatedNSRecordProps {
    /**
     * The hosted zone that requires the NS record. i.e domain.com
     */
    hostedZoneDomain: string;

    /**
     * The hosted zone to delegate. i.e. example.domain.com
     */
    delegatedDomainName: string;

    /**
     * A comma seperated list of name servers for the delegated hosted zone.
     * This can be derived from the NameServers output on the DelegatedHostedZone construct
     */
    nameServers: string;
}

/**
 * A Route 53 record set to delegate a hosted zone to another account
 */
export class DelegatedNSRecord extends Construct {
    constructor(scope: Construct, id: string, props: DelegatedNSRecordProps) {
        super(scope, id);

        const hostedZone = route53.HostedZone.fromLookup(this, 'HostedZone', {
            domainName: props.hostedZoneDomain
        });

        new route53.NsRecord(this, 'NSRecordSet', {
            zone: hostedZone,
            recordName: props.delegatedDomainName,
            values: props.nameServers.split(',')
        });
    }
}