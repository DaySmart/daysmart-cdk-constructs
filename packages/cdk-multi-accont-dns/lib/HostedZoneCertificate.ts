import * as cdk from 'aws-cdk-lib'; 
import * as acm from 'aws-cdk-lib/aws-certificatemanager';
import * as route53 from 'aws-cdk-lib/aws-route53';
import { Construct } from 'constructs';

/**
 * Properties for a ACM certificate for a hosted zone
 */
export interface HostedZoneCertificateProps {
    /**
     * The domain name of the hosted zone. i.e example.domain.com
     */
    domainName: string;
}

/**
 * An ACM certificate for all domains in a hosted zone
 * For example the domain example.domain.com would include:
 *  - example.domain.com
 *  - *.example.domain.com
 */
export class HostedZoneCertificate extends Construct {
    constructor(scope: Construct, id: string, props: HostedZoneCertificateProps) {
        super(scope, id);

        const hostedZone = route53.HostedZone.fromLookup(this, 'HostedZone', {
            domainName: props.domainName
        });

        const certificate = new acm.DnsValidatedCertificate(this, 'Certificate', {
            hostedZone: hostedZone,
            domainName: props.domainName,
            subjectAlternativeNames: [`*.${props.domainName}`]
        });

        const certificateArn = new cdk.CfnOutput(this, 'CertificateArn', {
            value: certificate.certificateArn
        });
        certificateArn.overrideLogicalId('CertificateArn');
    }
}