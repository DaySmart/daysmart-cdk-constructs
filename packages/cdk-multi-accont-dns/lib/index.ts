export { DelegatedHostedZone, DelegatedHostedZoneProps } from './DelegatedHostedZone';
export { DelegatedNSRecord, DelegatedNSRecordProps } from './DelegationNSRecord';
export { HostedZoneCertificate, HostedZoneCertificateProps } from './HostedZoneCertificate'; 