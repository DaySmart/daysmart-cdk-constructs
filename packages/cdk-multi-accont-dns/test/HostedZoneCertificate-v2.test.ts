import * as cdk from 'aws-cdk-lib';
import { Template } from "aws-cdk-lib/assertions"
import { HostedZoneCertificate } from '../lib/index'

test('App Cloudfront', () => {
    const stack = new cdk.Stack(undefined, 'stack', {
        env: {
            account: '123456',
            region: 'us-east-1'
        }
    });
    new HostedZoneCertificate(stack, 'AppCloudfront', {
        domainName: 'example.com',
    });

    const template = Template.fromStack(stack);
    expect(Template.fromStack(stack)).toMatchSnapshot();
})