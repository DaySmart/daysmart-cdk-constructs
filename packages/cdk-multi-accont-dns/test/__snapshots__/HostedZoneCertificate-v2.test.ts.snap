// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Cloudfront 1`] = `
Object {
  "Outputs": Object {
    "CertificateArn": Object {
      "Value": Object {
        "Ref": "AppCloudfrontCertificate69B623E7",
      },
    },
  },
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "AppCloudfrontCertificate69B623E7": Object {
      "Properties": Object {
        "DomainName": "example.com",
        "DomainValidationOptions": Array [
          Object {
            "DomainName": "example.com",
            "HostedZoneId": "DUMMY",
          },
        ],
        "SubjectAlternativeNames": Array [
          "*.example.com",
        ],
        "ValidationMethod": "DNS",
      },
      "Type": "AWS::CertificateManager::Certificate",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
